devito/types/sparse.py

```diff
@@ -111,20 +111,30 @@ def _support(self):
         The grid points surrounding each sparse point within the radius of self's
         injection/interpolation operators.
         """
+        return self._build_support(points=self.gridpoints)
+
+    @memoized_meth
+    def _build_support(self, points=None):
         ret = []
-        for i in self.gridpoints:
+        points = points or self.gridpoints
+        for i in points:
             support = [range(max(0, j - self._radius + 1), min(M, j + self._radius + 1))
                        for j, M in zip(i, self.grid.shape)]
             ret.append(tuple(product(*support)))
         return tuple(ret)
 
     @property
     def _dist_datamap(self):
+        return self._build_dist_datamap(support=self._support)
+
+    @memoized_meth
+    def _build_dist_datamap(self, support=None):
         """
         Mapper ``M : MPI rank -> required sparse data``.
         """
         ret = {}
-        for i, s in enumerate(self._support):
+        support = support or self._support
+        for i, s in enumerate(support):
             # Sparse point `i` is "required" by the following ranks
             for r in self.grid.distributor.glb_to_rank(s):
```

Comment:
@FabioLuporini that's the bad line overall, `_support` is fast and can be unmemoized if too overkill

Reply:
yeah, should only memoize the killer

also...is the actual culprit somewhere around here?
https://github.com/devitocodes/devito/blob/master/devito/mpi/distributed.py#L258

Reply:
Not sure, wasn't able to find exactly where it slows down.